### PR TITLE
[WebAssembly] Add SWIFT_ONCE_F implementation to Lazy.h

### PR DESCRIPTION
**Pull request by MaxDesiatov**
`SWIFT_ONCE_F` has to be implemented separately for WebAssembly/WASI as [WebAssembly currently doesn't support threads](https://bugs.swift.org/browse/SR-12097).

Related to SR-9307.

(cc @compnerd @kateinoigakukun)

### CODE REVIEW

File: include/swift/Basic/Lazy.h [hunk 1]
```diff
@@ -24,6 +24,22 @@
 #include "swift/Basic/Malloc.h"
 #include "swift/Basic/type_traits.h"
 
+#ifdef __wasi__
+inline void wasi_call_once(int *flag, void *context, void (*func)(void *)) {
+  switch (*flag) {
+    case 0:
```
Comment (xwu) [inline include/swift/Basic/Lazy.h:30]: Nit: LLVM style is to de-indent `case`.

**Comment by CodaFi** [comment]
My immediate concern is that we have documentation about how this implementation suffices but should be immediately replaced with one grounded in atomics when the WASM folk get around to merging support for concurrency.

Could you open an SR and leave behind a FIXME referencing that?

**Comment by MaxDesiatov** [comment]
@CodaFi there's already a reference to the corresponding SR-12097 at [the header of the file in `master`](https://github.com/apple/swift/blob/c70dce0226fbc059b154fea149760b49e849b4d1/include/swift/Basic/Lazy.h#L20), couple of lines above, not captured by the GitHub PR diff, unfortunately. Would you like it to be duplicated at the `#ifdef` too?

**Comment by CodaFi** [comment]
I’d like a separate issue about this function specifically.  The SR is OK as an umbrella, but there’s no signal in there about what is and isn’t done.

**Comment by MaxDesiatov** [comment]
Makes sense, that's done now.

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

File: include/swift/Basic/Lazy.h [hunk 2]
```diff
@@ -24,6 +24,24 @@
 #include "swift/Basic/Malloc.h"
 #include "swift/Basic/type_traits.h"
 
+#ifdef __wasi__
```
Comment (compnerd) [inline include/swift/Basic/Lazy.h:27]: Would you mind using `#if define(__wasi__)` please?

**Review by compnerd — APPROVED** [review]

**Comment by compnerd** [comment]
@swift-ci please test

**Pull request merged**